core/src/triggers.rs

```diff
@@ -0,0 +1,176 @@
+        self.hooks.contains_key(key)
+    }
+
+    /// Add more repetitions to the hook identified by [`trigger::Id`].
+    ///
+    /// # Errors
+    /// - if trigger not found.
+    /// - if addition to remaining current trigger repeats
+    /// overflows. Indefinitely repeating triggers always cause an
+    /// overflow.
+    pub fn mint_repeats(&self, key: trigger::Id, amount: u32) -> Result<(), smartcontracts::Error> {
+        if self.hooks.contains_key(&key) {
+            let mut action = self.hooks.get_mut(&key).ok_or(FindError::Trigger(key))?;
+            let new_repeats = match action.value().repeats {
+                Repeats::Exactly(n) => n
+                    .checked_add(amount)
+                    .ok_or(smartcontracts::Error::Math(MathError::Overflow)),
+                _ => Err(smartcontracts::Error::Math(MathError::Overflow)),
+            }?;
+            action.value_mut().repeats = Repeats::Exactly(new_repeats);
+            Ok(())
+        } else {
+            Err(smartcontracts::Error::Find(Box::new(FindError::Trigger(
+                key,
+            ))))
+        }
+    }
+
+    /// Reduce repetitions of the hook identified by [`trigger::Id`].
+    ///
+    /// # Errors
+    /// - if trigger not found.
+    /// - if subtraction from remaining current trigger repeats
+    /// underflows. Indefinitely repeating triggers always cause an
+    /// underflow.
+    pub fn burn_repeats(&self, key: trigger::Id, amount: u32) -> Result<(), smartcontracts::Error> {
+        if self.hooks.contains_key(&key) {
+            let mut action = self.hooks.get_mut(&key).ok_or(FindError::Trigger(key))?;
+            let new_repeats = match action.value().repeats {
+                Repeats::Indefinitely => Err(smartcontracts::Error::Math(MathError::Overflow)),
+                Repeats::Exactly(n) => n
+                    .checked_sub(amount)
+                    .ok_or(smartcontracts::Error::Math(MathError::Overflow)),
+            }?;
+            action.value_mut().repeats = Repeats::Exactly(new_repeats);
+            Ok(())
+        } else {
+            Err(smartcontracts::Error::Find(Box::new(FindError::Trigger(
+                key,
+            ))))
+        }
+    }
+
+    /// Produce and store recommendations for next block execution.
+    ///
+    /// # Panics
+    /// If locking recommendations for writing fails.
+    pub fn produce_recommendations(&self, events: &[Event]) {
+        let actions = self.actions_matching(events);
+        #[allow(clippy::expect_used)]
+        let mut recommendations = self
+            .recommendations
+            .write()
+            .expect("Failed to lock recommendations, when updating triggers.");
+        *recommendations = actions;
+    }
+
+    /// Find all actions which match the current events.
+    pub fn actions_matching(&self, events: &[Event]) -> Vec<Action> {
+        let mut result = Vec::new();
+        for event in events {
```

Comment:
This is quadratic, because `DashMap` is keyed in the trigger's name. 

A solution would be to keep a few `DashMaps` with specialised filters, and custom logic. I'd leave that for a much later stage if possible.